src/server/api/controllers/admin/EventsController.ts

```diff
@@ -141,12 +135,65 @@ export class EventsController {
   @UseBefore(RoleAuth(Role.ROLE_ADMIN))
   @UseBefore(ValidateEventID)
   async getTeams(@AuthorisedAdmin() admin: Admin, @SelectedEventID() event: EventDocument): Promise<TESCTeam[]> {
+    const isOrganiser = await this.EventService.isAdminOrganiser(event.alias, admin);
+    if (!isOrganiser) {
+      throw new ForbiddenError(ErrorMessage.PERMISSION_ERROR());
+    }
+
+    const teams = await this.TeamService.getTeamsByEvent(event);
+    return Promise.all(teams.map(this.TeamService.populateTeammatesAdminFields))
+  }
+
+  @Patch('/:eventId/teams')
+  @UseBefore(RoleAuth(Role.ROLE_ADMIN))
+  @UseBefore(ValidateEventID)
+  async updateTeam(
+    @AuthorisedAdmin() admin: Admin,
+    @SelectedEventID() event: EventDocument,
+    @Body() req: UpdateTeamRequest
+  ) {
     const isOrganiser = await this.EventService.isAdminOrganiser(event.alias, admin);
     if (!isOrganiser) {
       throw new BadRequestError(ErrorMessage.PERMISSION_ERROR());
     }
 
-    return this.TeamService.getTeamsByEvent(event);
+    const teamId = req.team._id;
+    if (!teamId) throw new BadRequestError('The team ID must be included in the request');
+    
+    let team: TeamDocument = await this.TeamService.getTeamById(teamId);
+    if (!team) throw new BadRequestError(`No team with ID ${teamId} found`);
+
+    // Add/Remove users based on patch body
+    const fetchUserByEmail = async email => {
+      const user = await this.UserService.getUserByEventAndEmail(email, event);
+      if (!user) throw new BadRequestError(ErrorMessage.NO_ACCOUNT_EXISTS());
+
+      return user;
+    };
+
+    const addUserAccounts = !!req.addMembers ? await Promise.all(req.addMembers.map(fetchUserByEmail)) : false;
+    const removeUserAccounts = !!req.removeMembers ? await Promise.all(req.removeMembers.map(fetchUserByEmail)) : false;
+
+    if (addUserAccounts) {
+      try {
+        await this.TeamService.addMembersToTeam(team, addUserAccounts);
+      } catch (e) {
+        throw new BadRequestError(e.message);
+      }
+    }
+    if (removeUserAccounts) {
+      try {
+        await this.TeamService.removeMembersToTeam(team, removeUserAccounts);
+      } catch (e) {
+        throw new BadRequestError(e.message);
```

Comment:
Should we not be removing before adding? I haven't run this myself but what happens if we have users A B C D and then edit to have users A X Y Z. Does this mean that we will make the team have 7 users before we remove B C D? sounds bad.

Reply:
We should probably be running these 2 operations in a transaction too

Reply:
Yeah in order to not hit the member limit we would have to remove before adding. This is a soft limit that we impose, but the `User` model requires a Team ID which can only be one value or another.
Transactions were added in Mongoose 5.2.0, so I'll try to add this in, too.